File: 3.EDPs/taller-poisson/poisson.f90
! poisson.f90 — Solución de la ecuación de Poisson con LAPACK dgbsv
!
!   ∇²V = (x²+y²)·exp(x·y)  en Ω = [0,2]×[0,1]
!   Solución analítica: V(x,y) = exp(x·y)
!
! Compilación:
!   gfortran -O2 poisson.f90 -o poisson_f -llapack -lblas
!
! Uso:
!   ./poisson_f N

program poisson_fd
    implicit none

    integer          :: N, argc
    character(len=32):: arg

    argc = command_argument_count()
    if (argc < 1) then
        write(*,*) "Uso: ./poisson_f N"
        stop
    end if
    call get_command_argument(1, arg)
    read(arg,*) N

    call solve_poisson(N)

end program poisson_fd

! ─────────────────────────────────────────────────────────────────────────────
subroutine solve_poisson(N)
    implicit none
    integer, intent(in) :: N

    ! Declaraciones explícitas de funciones externas
    double precision    :: f_rhs, V_exact, bc_left, bc_right, bc_bot, bc_top

    integer             :: M, sz, kl, ku, ldab
    integer             :: i, j, k, ki, kj, info
    double precision    :: hx, hy, ax, ay, ac
    double precision    :: x, y, diff, errL2, errMax, mem_MB
    double precision    :: t0, t1, elapsed_ms

    ! Variables para LAPACK dgbsv
    integer, allocatable          :: ipiv(:)
    double precision, allocatable :: AB(:,:), b_vec(:)

    ! Para métricas
    character(len=64)  :: fname_csv, fname_met

    M    = N
    sz   = N * M
    hx   = 2.0d0 / dble(N + 1)
    hy   = 1.0d0 / dble(N + 1)
    ax   = 1.0d0 / (hx*hx)
    ay   = 1.0d0 / (hy*hy)
    ac   = -2.0d0*ax - 2.0d0*ay

    ! Ancho de banda: kl = ku = N  (vecinos en y a distancia N)
    kl   = N
    ku   = N
    ldab = 2*kl + ku + 1

    allocate(AB(ldab, sz))
    allocate(b_vec(sz))
    allocate(ipiv(sz))
    AB     = 0.0d0
    b_vec  = 0.0d0

    ! Tiempo de inicio (cpu_time)
    call cpu_time(t0)

    ! ── Construcción de la matriz en formato banda LAPACK ──────────────────
    ! AB(kl+ku+1 + i - j, j) = A(i,j)   → offset = kl+ku+1
    do j = 0, M-1
        y = dble(j+1)*hy
        do i = 0, N-1
            x = dble(i+1)*hx
            k = j*N + i + 1          ! índice Fortran (1-base)

            ! diagonal
            AB(kl+ku+1, k) = ac

            ! vecino izquierdo (k-1)
            if (i > 0) then
                ki = k - 1
                AB(kl+ku+1 + k - ki, ki) = ax    ! AB(kl+ku+2, k-1)
            else
                b_vec(k) = b_vec(k) - ax * bc_left(y)
            end if

            ! vecino derecho (k+1)
            if (i < N-1) then
                ki = k + 1
                AB(kl+ku+1 + k - ki, ki) = ax    ! AB(kl+ku, k+1)
            else
                b_vec(k) = b_vec(k) - ax * bc_right(y)
            end if

            ! vecino inferior (k-N)
            if (j > 0) then
                kj = k - N
                AB(kl+ku+1 + k - kj, kj) = ay
            else
                b_vec(k) = b_vec(k) - ay * bc_bot(x)
            end if

            ! vecino superior (k+N)
            if (j < M-1) then
                kj = k + N
                AB(kl+ku+1 + k - kj, kj) = ay
            else
                b_vec(k) = b_vec(k) - ay * bc_top(x)
            end if

            ! fuente
            b_vec(k) = b_vec(k) + f_rhs(x, y)
        end do
    end do

    ! ── Resolver con LAPACK dgbsv ──────────────────────────────────────────
    call dgbsv(sz, kl, ku, 1, AB, ldab, ipiv, b_vec, sz, info)

    call cpu_time(t1)
    elapsed_ms = (t1 - t0) * 1.0d3

    if (info /= 0) then
        write(*,*) "[Fortran] Error en dgbsv: info =", info
        stop
    end if

    ! ── Errores ──────────────────────────────────────────────────────────────
    errL2  = 0.0d0
    errMax = 0.0d0
    do j = 0, M-1
        y = dble(j+1)*hy
        do i = 0, N-1
            x    = dble(i+1)*hx
            k    = j*N + i + 1
            diff = dabs(b_vec(k) - V_exact(x, y))
            errL2  = errL2 + diff*diff
            if (diff > errMax) errMax = diff
        end do
    end do
    errL2 = dsqrt(errL2 / dble(N*M))

    ! Memoria estimada: entradas de la banda con valores ~ 5*sz × 20 bytes
    mem_MB = dble(ldab * sz) * 8.0d0 / (1024.0d0*1024.0d0)

    write(*,'(A,I0,A,I0,A,I0,A)') "Fortran/LAPACK  malla ", N, "x", N, &
                                    "  (", N*N, " incognitas)"
    write(*,'(A,F12.4,A)') "  Tiempo    : ", elapsed_ms, " ms"
    write(*,'(A,F10.4,A)') "  Mem. est. : ", mem_MB, " MB"
    write(*,'(A,ES12.4)')  "  Error L2  : ", errL2
    write(*,'(A,ES12.4)')  "  Error Max : ", errMax

    ! ── CSV de la solución ────────────────────────────────────────────────────
    write(fname_csv,'(A,I0,A)') "sol_", N, "_fortran.csv"
    open(unit=10, file=trim(fname_csv), status="replace")
    write(10,'(A)') "x,y,V_num,V_exact"
    do j = 0, M-1
        y = dble(j+1)*hy
        do i = 0, N-1
            x = dble(i+1)*hx
            k = j*N + i + 1
            write(10,'(ES16.8,A,ES16.8,A,ES16.8,A,ES16.8)') &
                x,",",y,",",b_vec(k),",",V_exact(x,y)
        end do
    end do
    close(10)
    write(*,'(A,A)') "  CSV: ", trim(fname_csv)

    ! ── Métricas acumulativas ────────────────────────────────────────────────
    fname_met = "metrics_fortran.csv"
    ! Simplemente append; el notebook limpiará antes si es necesario
    open(unit=11, file=trim(fname_met), status="unknown", position="append")
    write(11,'(I0,A,I0,A,F12.4,A,F10.6,A,ES14.6,A,ES14.6)') &
        N,",",N*N,",",elapsed_ms,",",mem_MB,",",errL2,",",errMax
    close(11)

    deallocate(AB, b_vec, ipiv)

end subroutine solve_poisson

! ─────────────────────────────────────────────────────────────────────────────
double precision function f_rhs(x, y)
    double precision, intent(in) :: x, y
    f_rhs = (x*x + y*y) * dexp(x*y)
end function

double precision function V_exact(x, y)
    double precision, intent(in) :: x, y
    V_exact = dexp(x*y)
end function

double precision function bc_left(y)
    double precision, intent(in) :: y
    bc_left = 1.0d0
end function

double precision function bc_right(y)
    double precision, intent(in) :: y
    bc_right = dexp(2.0d0*y)
end function

double precision function bc_bot(x)
    double precision, intent(in) :: x
    bc_bot = 1.0d0
end function

double precision function bc_top(x)
    double precision, intent(in) :: x
    bc_top = dexp(x)
end function

File: 3.EDPs/taller-poisson/poisson.cpp
/*
 * poisson.cpp — Solución numérica de la ecuación de Poisson
 *
 *   ∇²V = (x² + y²)·exp(x·y)   en Ω = [0,2] × [0,1]
 *
 *   Condiciones de Dirichlet:
 *     V(0,y) = 1          V(2,y) = exp(2y)
 *     V(x,0) = 1          V(x,1) = exp(x)
 *
 *   Solución analítica: V(x,y) = exp(x·y)
 *
 * Compilación:
 *   g++ -O2 -std=c++17 -I/usr/include/eigen3 poisson.cpp -o poisson
 *
 * Uso:
 *   ./poisson N
 *     N : número de puntos interiores en cada dirección (e.g. 32)
 */

#include <iostream>
#include <fstream>
#include <sstream>
#include <cmath>
#include <chrono>
#include <string>
#include <vector>
#include <iomanip>

// ── Eigen (sparse) ──────────────────────────────────────────────────────────
#include <Eigen/Sparse>
#include <Eigen/SparseLU>
// ── Memoria real del proceso (VmRSS de /proc/self/status) ───────────────────
static double get_rss_mb()
{
    std::ifstream f("/proc/self/status");
    std::string line;
    while (std::getline(f, line))
    {
        if (line.rfind("VmRSS:", 0) == 0)
        {
            std::istringstream ss(line.substr(6));
            long kb;
            ss >> kb;
            return kb / 1024.0;
        }
    }
    return 0.0;
}
// ════════════════════════════════════════════════════════════════════════════
//  Funciones del problema
// ════════════════════════════════════════════════════════════════════════════

inline double f_rhs(double x, double y)
{
    return (x * x + y * y) * std::exp(x * y);
}

inline double V_exact(double x, double y)
{
    return std::exp(x * y);
}

// Condiciones de frontera
inline double bc_left(double y) { return 1.0; }
inline double bc_right(double y) { return std::exp(2.0 * y); }
inline double bc_bot(double x) { return 1.0; }
inline double bc_top(double x) { return std::exp(x); }

// ════════════════════════════════════════════════════════════════════════════
//  Construcción del sistema (compartida: mismos A e índices para ambos solvers)
// ════════════════════════════════════════════════════════════════════════════

struct GridData
{
    int N, M;
    double hx, hy;
};

inline int idx(int i, int j, int N) { return j * N + i; }

// ════════════════════════════════════════════════════════════════════════════
//  Eigen: construye y resuelve
// ════════════════════════════════════════════════════════════════════════════

std::vector<double> solve_eigen(const GridData &g,
                                double &elapsed_ms, double &mem_MB)
{
    const int N = g.N, M = g.M, sz = N * M;
    const double hx = g.hx, hy = g.hy;
    const double ax = 1.0 / (hx * hx), ay = 1.0 / (hy * hy);
    const double ac = -2.0 * ax - 2.0 * ay;

    using SpMat = Eigen::SparseMatrix<double>;
    using T = Eigen::Triplet<double>;

    auto t0 = std::chrono::high_resolution_clock::now();
    double mem_before = get_rss_mb(); // RSS antes de construir la matriz

    std::vector<T> triplets;
    triplets.reserve(5 * sz);
    Eigen::VectorXd b(sz);
    b.setZero();

    for (int j = 0; j < M; ++j)
    {
        double y = (j + 1) * hy;
        for (int i = 0; i < N; ++i)
        {
            double x = (i + 1) * hx;
            int k = idx(i, j, N);

            // diagonal
            triplets.emplace_back(k, k, ac);

            // vecino izquierdo
            if (i > 0)
                triplets.emplace_back(k, idx(i - 1, j, N), ax);
            else
                b[k] -= ax * bc_left(y);

            // vecino derecho
            if (i < N - 1)
                triplets.emplace_back(k, idx(i + 1, j, N), ax);
            else
                b[k] -= ax * bc_right(y);

            // vecino inferior
            if (j > 0)
                triplets.emplace_back(k, idx(i, j - 1, N), ay);
            else
                b[k] -= ay * bc_bot(x);

            // vecino superior
            if (j < M - 1)
                triplets.emplace_back(k, idx(i, j + 1, N), ay);
            else
                b[k] -= ay * bc_top(x);

            // término fuente
            b[k] += f_rhs(x, y);
        }
    }

    SpMat A(sz, sz);
    A.setFromTriplets(triplets.begin(), triplets.end());
    A.makeCompressed();

    Eigen::SparseLU<SpMat> solver;
    solver.analyzePattern(A);
    solver.factorize(A);
    if (solver.info() != Eigen::Success)
    {
        std::cerr << "[Eigen] Error en factorización.\n";
        return {};
    }
    Eigen::VectorXd u = solver.solve(b);

    auto t1 = std::chrono::high_resolution_clock::now();
    elapsed_ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    mem_MB = get_rss_mb() - mem_before; // delta RSS atribuible al solver

    return std::vector<double>(u.data(), u.data() + sz);
}

// ════════════════════════════════════════════════════════════════════════════
//  Métricas de error
// ════════════════════════════════════════════════════════════════════════════

void compute_errors(const std::vector<double> &u, const GridData &g,
                    double &errL2, double &errMax)
{
    const int N = g.N, M = g.M;
    errL2 = 0.0;
    errMax = 0.0;
    for (int j = 0; j < M; ++j)
    {
        double y = (j + 1) * g.hy;
        for (int i = 0; i < N; ++i)
        {
            double x = (i + 1) * g.hx;
            double diff = std::abs(u[idx(i, j, N)] - V_exact(x, y));
            errL2 += diff * diff;
            errMax = std::max(errMax, diff);
        }
    }
    errL2 = std::sqrt(errL2 / (N * M));
}

// ════════════════════════════════════════════════════════════════════════════
//  Guardar solución en CSV para visualización con Python
// ════════════════════════════════════════════════════════════════════════════

void save_csv(const std::string &fname,
              const std::vector<double> &u,
              const GridData &g)
{
    std::ofstream f(fname);
    f << std::scientific << std::setprecision(10);
    f << "x,y,V_num,V_exact\n";
    for (int j = 0; j < g.M; ++j)
    {
        double y = (j + 1) * g.hy;
        for (int i = 0; i < g.N; ++i)
        {
            double x = (i + 1) * g.hx;
            f << x << "," << y << ","
              << u[idx(i, j, g.N)] << ","
              << V_exact(x, y) << "\n";
        }
    }
}

// ════════════════════════════════════════════════════════════════════════════
//  Main
// ════════════════════════════════════════════════════════════════════════════

int main(int argc, char *argv[])
{
    if (argc < 2)
    {
        std::cerr << "Uso: " << argv[0] << " N\n"
                  << "  N : puntos interiores (e.g. 32)\n";
        return 1;
    }

    int N = std::atoi(argv[1]);
    if (N < 2)
    {
        std::cerr << "N debe ser >= 2\n";
        return 1;
    }

    GridData g;
    g.N = N;
    g.M = N;
    g.hx = 2.0 / (N + 1);
    g.hy = 1.0 / (N + 1);

    std::cout << std::fixed << std::setprecision(6);
    std::cout << "Poisson FD (Eigen)  malla " << N << "x" << N
              << "  (" << (long long)N * N << " incognitas)\n";

    double ms = 0, mem = 0;
    auto u = solve_eigen(g, ms, mem);
    if (u.empty())
        return 1;

    double eL2, eMax;
    compute_errors(u, g, eL2, eMax);

    std::cout << "  Tiempo    : " << ms << " ms\n"
              << "  Mem. RSS  : " << mem << " MB\n"
              << "  Error L2  : " << eL2 << "\n"
              << "  Error Max : " << eMax << "\n";

    // CSV de la solución
    std::string csvname = "sol_" + std::to_string(N) + "_eigen.csv";
    save_csv(csvname, u, g);
    std::cout << "  CSV: " << csvname << "\n";

    // Métricas acumulativas
    std::ofstream mf("metrics.csv", std::ios::app);
    if (mf.is_open())
    {
        mf.seekp(0, std::ios::end);
        if (mf.tellp() == 0)
            mf << "N,unknowns,time_ms,mem_MB,errL2,errMax\n";
        mf << N << "," << (long long)N * N << ","
           << ms << "," << mem << "," << eL2 << "," << eMax << "\n";
    }

    return 0;
}
